src/infercnvpy/tl/_infercnv.py

```diff
@@ -161,26 +181,83 @@ def _running_mean(x: np.ndarray | scipy.sparse.spmatrix, n: int = 50, step: int
         Length of the running window
     step
         only compute running windows ever `step` columns, e.g. if step is 10
-        0:100, 10:110, 20:120 etc. Saves memory.
+        0:99, 10:109, 20:119 etc. Saves memory.
     """
     if n < x.shape[1]:  # regular convolution: the filter is smaller than the #genes
         r = np.arange(1, n + 1)
-
         pyramid = np.minimum(r, r[::-1])
-        smoothed_x = np.apply_along_axis(lambda row: np.convolve(row, pyramid, mode="same"), axis=1, arr=x) / np.sum(
-            pyramid
-        )
-        return smoothed_x[:, np.arange(0, smoothed_x.shape[1], step)]
+        smoothed_x = np.apply_along_axis(
+            lambda row: np.convolve(row, pyramid, mode="valid"),
+            axis=1,
+            arr=x,
+        ) / np.sum(pyramid)
 
-    else:  # there's less genes than the filtersize. just apply a single conv with a smaller filter (no sliding)
-        r = np.arange(1, x.shape[1] + 1)
-        pyramid = np.minimum(r, r[::-1])
-        smoothed_x = x @ pyramid.reshape(-1, 1)
-        smoothed_x = smoothed_x / pyramid.sum()
-        return smoothed_x
+        ## get the indices of the genes used in the convolution
+        convolution_indices = get_convolution_indices(x, n)[np.arange(0, smoothed_x.shape[1], step)]
+        ## Pull out the genes used in the convolution
+        convolved_gene_names = gene_list[convolution_indices]
+        smoothed_x = smoothed_x[:, np.arange(0, smoothed_x.shape[1], step)]
 
+        convolved_gene_values = _calculate_gene_averages(convolved_gene_names, smoothed_x)
 
-def _running_mean_by_chromosome(expr, var, window_size, step) -> tuple[dict, np.ndarray]:
+        return smoothed_x, convolved_gene_values
+
+    else:  # If there is less genes than the window size, set the window size to the number of genes and perform a single convolution
+        n = x.shape[1]  # set the filter size to the number of genes
+        r = np.arange(1, n + 1)
+        ## As we are only doing one convolution the values should be equal
+        pyramid = np.array([1] * n)
+        # Old code to make the pyramid
+        # pyramid = np.minimum(r, r[::-1])
+        smoothed_x = np.apply_along_axis(
+            lambda row: np.convolve(row, pyramid, mode="valid"),
+            axis=1,
+            arr=x,
+        ) / np.sum(pyramid)
+
+        ## As all genes are used the convolution the values are identical for all genes
+        convolved_gene_values = pd.DataFrame(np.repeat(smoothed_x, len(gene_list), axis=1), columns=gene_list)
+
+        return smoothed_x, convolved_gene_values
```

Comment:
Now the if and else block are even more similar than before. Do you think we could get rid of the code duplication?